include/cuco/detail/bloom_filter/bloom_filter_impl.cuh

```diff
@@ -0,0 +1,359 @@
+                     OutputIt output_begin,
+                     cuda::stream_ref stream) const
+  {
+    this->test_async(first, last, output_begin, stream);
+    stream.wait();
+  }
+
+  template <class InputIt, class OutputIt>
+  __host__ void test_async(InputIt first,
+                           InputIt last,
+                           OutputIt output_begin,
+                           cuda::stream_ref stream) const noexcept
+  {
+    auto const always_true = thrust::constant_iterator<bool>{true};
+    this->test_if_async(first, last, always_true, thrust::identity{}, output_begin, stream);
+  }
+
+  template <class InputIt, class StencilIt, class Predicate, class OutputIt>
+  __host__ void test_if(InputIt first,
+                        InputIt last,
+                        StencilIt stencil,
+                        Predicate pred,
+                        OutputIt output_begin,
+                        cuda::stream_ref stream) const
+  {
+    this->test_if_async(first, last, stencil, pred, output_begin, stream);
+    stream.wait();
+  }
+
+  template <class InputIt, class StencilIt, class Predicate, class OutputIt>
+  __host__ void test_if_async(InputIt first,
+                              InputIt last,
+                              StencilIt stencil,
+                              Predicate pred,
+                              OutputIt output_begin,
+                              cuda::stream_ref stream) const noexcept
+  {
+    auto const num_keys = cuco::detail::distance(first, last);
+    if (num_keys == 0) { return; }
+
+    auto constexpr block_size = cuco::detail::default_block_size();
+    auto const grid_size =
+      cuco::detail::grid_size(num_keys, 1, cuco::detail::default_stride(), block_size);
+
+    detail::test_if_n<block_size><<<grid_size, block_size, 0, stream.get()>>>(
+      first, num_keys, stencil, pred, output_begin, *this);
+  }
+
+  [[nodiscard]] __host__ __device__ word_type* data() noexcept { return words_; }
+
+  [[nodiscard]] __host__ __device__ word_type const* data() const noexcept { return words_; }
+
+  [[nodiscard]] __host__ __device__ extent_type block_extent() const noexcept
+  {
+    return num_blocks_;
+  }
+
+  [[nodiscard]] __host__ __device__ uint32_t pattern_bits() const noexcept { return pattern_bits_; }
+
+  [[nodiscard]] __host__ __device__ hasher hash_function() const noexcept { return hash_; }
+
+  // TODO
+  // [[nodiscard]] __host__ double occupancy() const;
+  // [[nodiscard]] __host__ double expected_false_positive_rate(size_t unique_keys) const
+  // [[nodiscard]] __host__ __device__ static uint32_t optimal_pattern_bits(size_t num_blocks)
+  // template <typename CG, cuda::thread_scope NewScope = thread_scope>
+  // [[nodiscard]] __device__ constexpr auto make_copy(CG const& group, word_type* const
+  // memory_to_use, cuda_thread_scope<NewScope> scope = {}) const noexcept;
+
+ private:
+  template <class HashValue>
+  __device__ size_type block_idx(HashValue hash_value) const
```

Comment:
IIRC `std::hash::result_type` is deprecated since C++17 so we would have to jump through some hoops with `decltype/declval` to get the actual result type of the hasher. I thought a simple tparam would solve this problem with less lines of code.

Reply:
It's also a matter of build time: while adding extra templates for convenience simplifies our development, it increases build time for users, causing unnecessary delays.